File: react-native/src/Telemetry.ts
/**
 * Edge Veda SDK - Telemetry
 * Structured logging and metrics collection for runtime supervision.
 *
 * Provides console-based logging with appropriate categories and stores
 * performance metrics (latency, budget violations, resource snapshots)
 * for retrieval and analysis.
 *
 * JavaScript is single-threaded in React Native, so no mutex is needed.
 *
 * @example
 * ```typescript
 * const telemetry = Telemetry.instance;
 * telemetry.logInferenceStart('req-123', 'my-model');
 * telemetry.recordLatency('req-123', 42.5);
 * telemetry.logBudgetViolation(BudgetViolationType.LATENCY, 50, 40);
 * ```
 */

// =============================================================================
// Supporting Types
// =============================================================================

/** Budget violation type. */
export enum BudgetViolationType {
  LATENCY = 'latency',
  MEMORY = 'memory',
  BATTERY = 'battery',
  THERMAL = 'thermal',
}

/** Violation severity level. */
export enum ViolationSeverity {
  INFO = 'info',
  WARNING = 'warning',
  CRITICAL = 'critical',
}

/** Latency metric for a single inference request. */
export interface LatencyMetric {
  requestId: string;
  modelName?: string;
  startTime?: Date;
  endTime?: Date;
  latencyMs?: number;
}

/** Budget violation record. */
export interface BudgetViolationRecord {
  timestamp: Date;
  type: BudgetViolationType;
  current: number;
  limit: number;
  severity: ViolationSeverity;
}

/** Resource usage snapshot. */
export interface ResourceSnapshot {
  timestamp: Date;
  memoryMb: number;
  batteryLevel?: number;
  thermalLevel: number;
}

/** Aggregated latency statistics. */
export interface LatencyStats {
  count: number;
  min: number;
  max: number;
  mean: number;
  p50: number;
  p95: number;
  p99: number;
}

/** Format LatencyStats as a human-readable string. */
export function latencyStatsToString(s: LatencyStats): string {
  return (
    `LatencyStats(count=${s.count}, min=${s.min.toFixed(2)}ms, max=${s.max.toFixed(2)}ms, ` +
    `mean=${s.mean.toFixed(2)}ms, p50=${s.p50.toFixed(2)}ms, ` +
    `p95=${s.p95.toFixed(2)}ms, p99=${s.p99.toFixed(2)}ms)`
  );
}

// =============================================================================
// Log Tags
// =============================================================================

const TAG_INFERENCE = 'EdgeVeda.Inference';
const TAG_BUDGET = 'EdgeVeda.Budget';
const TAG_RESOURCES = 'EdgeVeda.Resources';
const TAG_SCHEDULER = 'EdgeVeda.Scheduler';
const TAG_POLICY = 'EdgeVeda.Policy';

// =============================================================================
// Telemetry
// =============================================================================

/**
 * Singleton telemetry subsystem for structured logging and metrics collection.
 *
 * Access via `Telemetry.instance`.
 */
export class Telemetry {
  // -------------------------------------------------------------------------
  // Singleton
  // -------------------------------------------------------------------------

  private static _instance: Telemetry | null = null;

  static get instance(): Telemetry {
    if (!Telemetry._instance) {
      Telemetry._instance = new Telemetry();
    }
    return Telemetry._instance;
  }

  private constructor() {
    console.info(`[${TAG_INFERENCE}] Telemetry system initialized`);
  }

  // -------------------------------------------------------------------------
  // Metrics Storage
  // -------------------------------------------------------------------------

  private latencyMetrics = new Map<string, LatencyMetric>();
  private budgetViolations: BudgetViolationRecord[] = [];
  private resourceSnapshots: ResourceSnapshot[] = [];

  private maxStoredMetrics = 1000;
  private maxStoredViolations = 100;
  private maxStoredSnapshots = 100;

  // -------------------------------------------------------------------------
  // Configuration
  // -------------------------------------------------------------------------

  /** Set maximum number of latency metrics to store in memory. */
  setMaxStoredMetrics(count: number): void {
    this.maxStoredMetrics = count;
    this.trimMetrics();
  }

  /** Set maximum number of budget violations to store. */
  setMaxStoredViolations(count: number): void {
    this.maxStoredViolations = count;
    this.trimViolations();
  }

  /** Set maximum number of resource snapshots to store. */
  setMaxStoredSnapshots(count: number): void {
    this.maxStoredSnapshots = count;
    this.trimSnapshots();
  }

  // -------------------------------------------------------------------------
  // Inference Logging
  // -------------------------------------------------------------------------

  /** Log the start of an inference request. */
  logInferenceStart(requestId: string, modelName?: string): void {
    const msg = modelName
      ? `Inference started: ${requestId} model=${modelName}`
      : `Inference started: ${requestId}`;
    console.info(`[${TAG_INFERENCE}] ${msg}`);

    this.latencyMetrics.set(requestId, {
      requestId,
      modelName,
      startTime: new Date(),
    });
    this.trimMetrics();
  }

  /** Log the completion of an inference request. */
  logInferenceComplete(requestId: string, tokensGenerated?: number): void {
    const msg = tokensGenerated !== undefined
      ? `Inference completed: ${requestId} tokens=${tokensGenerated}`
      : `Inference completed: ${requestId}`;
    console.info(`[${TAG_INFERENCE}] ${msg}`);

    const metric = this.latencyMetrics.get(requestId);
    if (metric) {
      const endTime = new Date();
      const latencyMs = metric.startTime
        ? endTime.getTime() - metric.startTime.getTime()
        : undefined;
      this.latencyMetrics.set(requestId, {
        ...metric,
        endTime,
        latencyMs,
      });
    }
  }

  /** Log an inference error. */
  logInferenceError(requestId: string, error: string): void {
    console.error(`[${TAG_INFERENCE}] Inference failed: ${requestId} error=${error}`);
  }

  /** Record latency for a completed request. */
  recordLatency(requestId: string, latencyMs: number): void {
    console.debug(`[${TAG_INFERENCE}] Latency recorded: ${requestId} latency=${latencyMs}ms`);

    const metric = this.latencyMetrics.get(requestId);
    if (metric) {
      this.latencyMetrics.set(requestId, { ...metric, latencyMs });
    }
  }

  // -------------------------------------------------------------------------
  // Budget Logging
  // -------------------------------------------------------------------------

  /** Log a budget violation. */
  logBudgetViolation(
    type: BudgetViolationType,
    current: number,
    limit: number,
    severity: ViolationSeverity = ViolationSeverity.WARNING
  ): void {
    const message = `Budget violation: ${type} current=${current} limit=${limit} severity=${severity}`;

    switch (severity) {
      case ViolationSeverity.INFO:
        console.info(`[${TAG_BUDGET}] ${message}`);
        break;
      case ViolationSeverity.WARNING:
        console.warn(`[${TAG_BUDGET}] ${message}`);
        break;
      case ViolationSeverity.CRITICAL:
        console.error(`[${TAG_BUDGET}] ${message}`);
        break;
    }

    this.budgetViolations.push({
      timestamp: new Date(),
      type,
      current,
      limit,
      severity,
    });
    this.trimViolations();
  }

  /** Log budget enforcement action. */
  logBudgetEnforcement(action: string, reason: string): void {
    console.info(`[${TAG_BUDGET}] Budget enforcement: ${action} reason=${reason}`);
  }

  /** Log measured baseline update. */
  logBaselineUpdate(p50: number, p95: number, p99: number, sampleCount: number): void {
    console.info(
      `[${TAG_BUDGET}] Baseline updated: p50=${p50}ms p95=${p95}ms p99=${p99}ms samples=${sampleCount}`
    );
  }

  // -------------------------------------------------------------------------
  // Resource Logging
  // -------------------------------------------------------------------------

  /** Log current resource usage. */
  logResourceUsage(memoryMb: number, batteryLevel?: number, thermalLevel: number = 0): void {
    const msg = batteryLevel !== undefined
      ? `Resources: memory=${memoryMb}MB battery=${Math.round(batteryLevel * 100)}% thermal=${thermalLevel}`
      : `Resources: memory=${memoryMb}MB thermal=${thermalLevel}`;
    console.debug(`[${TAG_RESOURCES}] ${msg}`);

    this.resourceSnapshots.push({
      timestamp: new Date(),
      memoryMb,
      batteryLevel,
      thermalLevel,
    });
    this.trimSnapshots();
  }

  /** Log memory pressure event. */
  logMemoryPressure(current: number, peak: number, available?: number): void {
    const msg = available !== undefined
      ? `Memory pressure: current=${current}MB peak=${peak}MB available=${available}MB`
      : `Memory pressure: current=${current}MB peak=${peak}MB`;
    console.warn(`[${TAG_RESOURCES}] ${msg}`);
  }

  /** Log thermal state change. */
  logThermalStateChange(from: number, to: number): void {
    console.info(`[${TAG_RESOURCES}] Thermal state changed: ${from} -> ${to}`);
  }

  /** Log battery drain rate. */
  logBatteryDrain(drainRate: number, currentLevel: number): void {
    console.info(
      `[${TAG_RESOURCES}] Battery drain: rate=${drainRate}%/10min level=${Math.round(currentLevel * 100)}%`
    );
  }

  // -------------------------------------------------------------------------
  // Scheduler Logging
  // -------------------------------------------------------------------------

  /** Log task scheduling. */
  logTaskScheduled(taskId: string, priority: string): void {
    console.debug(`[${TAG_SCHEDULER}] Task scheduled: ${taskId} priority=${priority}`);
  }

  /** Log task execution start. */
  logTaskStarted(taskId: string): void {
    console.debug(`[${TAG_SCHEDULER}] Task started: ${taskId}`);
  }

  /** Log task completion. */
  logTaskCompleted(taskId: string, durationMs: number): void {
    console.debug(`[${TAG_SCHEDULER}] Task completed: ${taskId} duration=${durationMs}ms`);
  }

  /** Log task cancellation. */
  logTaskCancelled(taskId: string, reason: string): void {
    console.info(`[${TAG_SCHEDULER}] Task cancelled: ${taskId} reason=${reason}`);
  }

  /** Log queue status. */
  logQueueStatus(pending: number, running: number, priority: string): void {
    console.debug(
      `[${TAG_SCHEDULER}] Queue status: pending=${pending} running=${running} priority=${priority}`
    );
  }

  // -------------------------------------------------------------------------
  // Policy Logging
  // -------------------------------------------------------------------------

  /** Log policy change. */
  logPolicyChange(from: string, to: string): void {
    console.info(`[${TAG_POLICY}] Policy changed: ${from} -> ${to}`);
  }

  /** Log throttle decision. */
  logThrottleDecision(shouldThrottle: boolean, factor: number, reasons: string[]): void {
    if (shouldThrottle) {
      console.warn(
        `[${TAG_POLICY}] Throttling applied: factor=${factor} reasons=${reasons.join(', ')}`
      );
    } else {
      console.debug(`[${TAG_POLICY}] No throttling needed`);
    }
  }

  /** Log policy enforcement action. */
  logPolicyEnforcement(action: string, context: string): void {
    console.info(`[${TAG_POLICY}] Policy enforcement: ${action} context=${context}`);
  }

  // -------------------------------------------------------------------------
  // Metrics Retrieval
  // -------------------------------------------------------------------------

  /** Get all stored latency metrics. */
  getLatencyMetrics(): LatencyMetric[] {
    return Array.from(this.latencyMetrics.values());
  }

  /** Get latency metric for a specific request. */
  getLatencyMetric(requestId: string): LatencyMetric | undefined {
    return this.latencyMetrics.get(requestId);
  }

  /** Get all budget violations. */
  getBudgetViolations(): BudgetViolationRecord[] {
    return [...this.budgetViolations];
  }

  /** Get recent budget violations (last N). */
  getRecentViolations(count: number): BudgetViolationRecord[] {
    const start = Math.max(0, this.budgetViolations.length - count);
    return this.budgetViolations.slice(start);
  }

  /** Get all resource snapshots. */
  getResourceSnapshots(): ResourceSnapshot[] {
    return [...this.resourceSnapshots];
  }

  /** Get recent resource snapshots (last N). */
  getRecentSnapshots(count: number): ResourceSnapshot[] {
    const start = Math.max(0, this.resourceSnapshots.length - count);
    return this.resourceSnapshots.slice(start);
  }

  /** Get aggregated latency statistics, or undefined if no completed metrics. */
  getLatencyStats(): LatencyStats | undefined {
    const latencies: number[] = [];
    for (const metric of this.latencyMetrics.values()) {
      if (metric.latencyMs !== undefined) {
        latencies.push(metric.latencyMs);
      }
    }
    if (latencies.length === 0) return undefined;

    const sorted = [...latencies].sort((a, b) => a - b);
    const count = sorted.length;
    const sum = sorted.reduce((a, b) => a + b, 0);

    return {
      count,
      min: sorted[0]!,
      max: sorted[count - 1]!,
      mean: sum / count,
      p50: sorted[Math.min(Math.floor(count * 0.5), count - 1)]!,
      p95: sorted[Math.min(Math.floor(count * 0.95), count - 1)]!,
      p99: sorted[Math.min(Math.floor(count * 0.99), count - 1)]!,
    };
  }

  // -------------------------------------------------------------------------
  // Cleanup
  // -------------------------------------------------------------------------

  /** Clear all stored metrics. */
  clearMetrics(): void {
    this.latencyMetrics.clear();
    this.budgetViolations = [];
    this.resourceSnapshots = [];
    console.info(`[${TAG_INFERENCE}] All metrics cleared`);
  }

  /** Clear metrics older than specified duration in milliseconds. */
  clearOldMetrics(olderThanMs: number): void {
    const cutoff = new Date(Date.now() - olderThanMs);

    for (const [key, metric] of this.latencyMetrics) {
      if (!metric.startTime || metric.startTime < cutoff) {
        this.latencyMetrics.delete(key);
      }
    }

    this.budgetViolations = this.budgetViolations.filter(
      (v) => v.timestamp >= cutoff
    );
    this.resourceSnapshots = this.resourceSnapshots.filter(
      (s) => s.timestamp >= cutoff
    );

    console.info(`[${TAG_INFERENCE}] Cleared metrics older than ${olderThanMs}ms`);
  }

  // -------------------------------------------------------------------------
  // Private - Trimming
  // -------------------------------------------------------------------------

  private trimMetrics(): void {
    if (this.latencyMetrics.size > this.maxStoredMetrics) {
      // Remove oldest entries (Map preserves insertion order)
      const toRemove = this.latencyMetrics.size - this.maxStoredMetrics;
      let removed = 0;
      for (const key of this.latencyMetrics.keys()) {
        if (removed >= toRemove) break;
        this.latencyMetrics.delete(key);
        removed++;
      }
    }
  }

  private trimViolations(): void {
    if (this.budgetViolations.length > this.maxStoredViolations) {
      const toRemove = this.budgetViolations.length - this.maxStoredViolations;
      this.budgetViolations.splice(0, toRemove);
    }
  }

  private trimSnapshots(): void {
    if (this.resourceSnapshots.length > this.maxStoredSnapshots) {
      const toRemove = this.resourceSnapshots.length - this.maxStoredSnapshots;
      this.resourceSnapshots.splice(0, toRemove);
    }
  }
}

File: kotlin/src/main/kotlin/com/edgeveda/sdk/Telemetry.kt
package com.edgeveda.sdk

import android.util.Log
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.Date

/**
 * Telemetry subsystem for structured logging and metrics collection.
 *
 * Provides [android.util.Log]-based logging with appropriate categories.
 * Integrates with runtime supervision components to collect performance metrics.
 *
 * Thread-safe via [Mutex] (Kotlin equivalent of Swift actor isolation).
 *
 * Example:
 * ```kotlin
 * val telemetry = Telemetry.instance
 * telemetry.logInferenceStart(requestId = "req-123")
 * telemetry.recordLatency(requestId = "req-123", latencyMs = 42.5)
 * telemetry.logBudgetViolation(type = BudgetViolationType.LATENCY, current = 50.0, limit = 40.0)
 * ```
 */
class Telemetry private constructor() {

    private val mutex = Mutex()

    // -------------------------------------------------------------------
    // Metrics Storage
    // -------------------------------------------------------------------

    private val latencyMetrics = mutableMapOf<String, LatencyMetric>()
    private val budgetViolations = mutableListOf<BudgetViolationRecord>()
    private val resourceSnapshots = mutableListOf<ResourceSnapshot>()

    // Configuration
    private var maxStoredMetrics = 1000
    private var maxStoredViolations = 100
    private var maxStoredSnapshots = 100

    companion object {
        // Log tags per category
        private const val TAG_INFERENCE = "EdgeVeda.Inference"
        private const val TAG_BUDGET = "EdgeVeda.Budget"
        private const val TAG_RESOURCES = "EdgeVeda.Resources"
        private const val TAG_SCHEDULER = "EdgeVeda.Scheduler"
        private const val TAG_POLICY = "EdgeVeda.Policy"

        /** Shared telemetry instance (singleton). */
        val instance: Telemetry by lazy { Telemetry() }
    }

    init {
        Log.i(TAG_INFERENCE, "Telemetry system initialized")
    }

    // -------------------------------------------------------------------
    // Configuration
    // -------------------------------------------------------------------

    /** Set maximum number of latency metrics to store in memory. */
    suspend fun setMaxStoredMetrics(count: Int) = mutex.withLock {
        maxStoredMetrics = count
        trimMetrics()
    }

    /** Set maximum number of budget violations to store. */
    suspend fun setMaxStoredViolations(count: Int) = mutex.withLock {
        maxStoredViolations = count
        trimViolations()
    }

    /** Set maximum number of resource snapshots to store. */
    suspend fun setMaxStoredSnapshots(count: Int) = mutex.withLock {
        maxStoredSnapshots = count
        trimSnapshots()
    }

    // -------------------------------------------------------------------
    // Inference Logging
    // -------------------------------------------------------------------

    /** Log the start of an inference request. */
    suspend fun logInferenceStart(requestId: String, modelName: String? = null) {
        if (modelName != null) {
            Log.i(TAG_INFERENCE, "Inference started: $requestId model=$modelName")
        } else {
            Log.i(TAG_INFERENCE, "Inference started: $requestId")
        }

        val metric = LatencyMetric(
            requestId = requestId,
            modelName = modelName,
            startTime = Date(),
            endTime = null,
            latencyMs = null
        )

        mutex.withLock {
            latencyMetrics[requestId] = metric
            trimMetrics()
        }
    }

    /** Log the completion of an inference request. */
    suspend fun logInferenceComplete(requestId: String, tokensGenerated: Int? = null) {
        if (tokensGenerated != null) {
            Log.i(TAG_INFERENCE, "Inference completed: $requestId tokens=$tokensGenerated")
        } else {
            Log.i(TAG_INFERENCE, "Inference completed: $requestId")
        }

        mutex.withLock {
            latencyMetrics[requestId]?.let { metric ->
                val endTime = Date()
                val latencyMs = metric.startTime?.let { start ->
                    (endTime.time - start.time).toDouble()
                }
                latencyMetrics[requestId] = metric.copy(
                    endTime = endTime,
                    latencyMs = latencyMs
                )
            }
        }
    }

    /** Log an inference error. */
    fun logInferenceError(requestId: String, error: String) {
        Log.e(TAG_INFERENCE, "Inference failed: $requestId error=$error")
    }

    /** Record latency for a completed request. */
    suspend fun recordLatency(requestId: String, latencyMs: Double) {
        Log.d(TAG_INFERENCE, "Latency recorded: $requestId latency=${latencyMs}ms")

        mutex.withLock {
            latencyMetrics[requestId]?.let { metric ->
                latencyMetrics[requestId] = metric.copy(latencyMs = latencyMs)
            }
        }
    }

    // -------------------------------------------------------------------
    // Budget Logging
    // -------------------------------------------------------------------

    /** Log a budget violation. */
    suspend fun logBudgetViolation(
        type: BudgetViolationType,
        current: Double,
        limit: Double,
        severity: ViolationSeverity = ViolationSeverity.WARNING
    ) {
        val message = "Budget violation: ${type.value} current=$current limit=$limit severity=${severity.value}"

        when (severity) {
            ViolationSeverity.INFO -> Log.i(TAG_BUDGET, message)
            ViolationSeverity.WARNING -> Log.w(TAG_BUDGET, message)
            ViolationSeverity.CRITICAL -> Log.e(TAG_BUDGET, message)
        }

        val record = BudgetViolationRecord(
            timestamp = Date(),
            type = type,
            current = current,
            limit = limit,
            severity = severity
        )

        mutex.withLock {
            budgetViolations.add(record)
            trimViolations()
        }
    }

    /** Log budget enforcement action. */
    fun logBudgetEnforcement(action: String, reason: String) {
        Log.i(TAG_BUDGET, "Budget enforcement: $action reason=$reason")
    }

    /** Log measured baseline update. */
    fun logBaselineUpdate(p50: Double, p95: Double, p99: Double, sampleCount: Int) {
        Log.i(TAG_BUDGET, "Baseline updated: p50=${p50}ms p95=${p95}ms p99=${p99}ms samples=$sampleCount")
    }

    // -------------------------------------------------------------------
    // Resource Logging
    // -------------------------------------------------------------------

    /** Log current resource usage. */
    suspend fun logResourceUsage(
        memoryMb: Double,
        batteryLevel: Double? = null,
        thermalLevel: Int = 0
    ) {
        if (batteryLevel != null) {
            Log.d(TAG_RESOURCES, "Resources: memory=${memoryMb}MB battery=${(batteryLevel * 100).toInt()}% thermal=$thermalLevel")
        } else {
            Log.d(TAG_RESOURCES, "Resources: memory=${memoryMb}MB thermal=$thermalLevel")
        }

        val snapshot = ResourceSnapshot(
            timestamp = Date(),
            memoryMb = memoryMb,
            batteryLevel = batteryLevel,
            thermalLevel = thermalLevel
        )

        mutex.withLock {
            resourceSnapshots.add(snapshot)
            trimSnapshots()
        }
    }

    /** Log memory pressure event. */
    fun logMemoryPressure(current: Double, peak: Double, available: Double? = null) {
        if (available != null) {
            Log.w(TAG_RESOURCES, "Memory pressure: current=${current}MB peak=${peak}MB available=${available}MB")
        } else {
            Log.w(TAG_RESOURCES, "Memory pressure: current=${current}MB peak=${peak}MB")
        }
    }

    /** Log thermal state change. */
    fun logThermalStateChange(from: Int, to: Int) {
        Log.i(TAG_RESOURCES, "Thermal state changed: $from -> $to")
    }

    /** Log battery drain rate. */
    fun logBatteryDrain(drainRate: Double, currentLevel: Double) {
        Log.i(TAG_RESOURCES, "Battery drain: rate=${drainRate}%/10min level=${(currentLevel * 100).toInt()}%")
    }

    // -------------------------------------------------------------------
    // Scheduler Logging
    // -------------------------------------------------------------------

    /** Log task scheduling. */
    fun logTaskScheduled(taskId: String, priority: String) {
        Log.d(TAG_SCHEDULER, "Task scheduled: $taskId priority=$priority")
    }

    /** Log task execution start. */
    fun logTaskStarted(taskId: String) {
        Log.d(TAG_SCHEDULER, "Task started: $taskId")
    }

    /** Log task completion. */
    fun logTaskCompleted(taskId: String, durationMs: Double) {
        Log.d(TAG_SCHEDULER, "Task completed: $taskId duration=${durationMs}ms")
    }

    /** Log task cancellation. */
    fun logTaskCancelled(taskId: String, reason: String) {
        Log.i(TAG_SCHEDULER, "Task cancelled: $taskId reason=$reason")
    }

    /** Log queue status. */
    fun logQueueStatus(pending: Int, running: Int, priority: String) {
        Log.d(TAG_SCHEDULER, "Queue status: pending=$pending running=$running priority=$priority")
    }

    // -------------------------------------------------------------------
    // Policy Logging
    // -------------------------------------------------------------------

    /** Log policy change. */
    fun logPolicyChange(from: String, to: String) {
        Log.i(TAG_POLICY, "Policy changed: $from -> $to")
    }

    /** Log throttle decision. */
    fun logThrottleDecision(shouldThrottle: Boolean, factor: Double, reasons: List<String>) {
        if (shouldThrottle) {
            Log.w(TAG_POLICY, "Throttling applied: factor=$factor reasons=${reasons.joinToString(", ")}")
        } else {
            Log.d(TAG_POLICY, "No throttling needed")
        }
    }

    /** Log policy enforcement action. */
    fun logPolicyEnforcement(action: String, context: String) {
        Log.i(TAG_POLICY, "Policy enforcement: $action context=$context")
    }

    // -------------------------------------------------------------------
    // Metrics Retrieval
    // -------------------------------------------------------------------

    /** Get all stored latency metrics. */
    suspend fun getLatencyMetrics(): List<LatencyMetric> = mutex.withLock {
        latencyMetrics.values.toList()
    }

    /** Get latency metric for a specific request. */
    suspend fun getLatencyMetric(requestId: String): LatencyMetric? = mutex.withLock {
        latencyMetrics[requestId]
    }

    /** Get all budget violations. */
    suspend fun getBudgetViolations(): List<BudgetViolationRecord> = mutex.withLock {
        budgetViolations.toList()
    }

    /** Get recent budget violations (last N). */
    suspend fun getRecentViolations(count: Int): List<BudgetViolationRecord> = mutex.withLock {
        val startIndex = maxOf(0, budgetViolations.size - count)
        budgetViolations.subList(startIndex, budgetViolations.size).toList()
    }

    /** Get all resource snapshots. */
    suspend fun getResourceSnapshots(): List<ResourceSnapshot> = mutex.withLock {
        resourceSnapshots.toList()
    }

    /** Get recent resource snapshots (last N). */
    suspend fun getRecentSnapshots(count: Int): List<ResourceSnapshot> = mutex.withLock {
        val startIndex = maxOf(0, resourceSnapshots.size - count)
        resourceSnapshots.subList(startIndex, resourceSnapshots.size).toList()
    }

    /** Get aggregated latency statistics, or null if no completed metrics. */
    suspend fun getLatencyStats(): LatencyStats? = mutex.withLock {
        val latencies = latencyMetrics.values.mapNotNull { it.latencyMs }
        if (latencies.isEmpty()) return@withLock null

        val sorted = latencies.sorted()
        val count = sorted.size.toDouble()

        LatencyStats(
            count = sorted.size,
            min = sorted.first(),
            max = sorted.last(),
            mean = sorted.sum() / count,
            p50 = sorted[(count * 0.5).toInt().coerceAtMost(sorted.size - 1)],
            p95 = sorted[(count * 0.95).toInt().coerceAtMost(sorted.size - 1)],
            p99 = sorted[(count * 0.99).toInt().coerceAtMost(sorted.size - 1)]
        )
    }

    // -------------------------------------------------------------------
    // Cleanup
    // -------------------------------------------------------------------

    /** Clear all stored metrics. */
    suspend fun clearMetrics() = mutex.withLock {
        latencyMetrics.clear()
        budgetViolations.clear()
        resourceSnapshots.clear()
        Log.i(TAG_INFERENCE, "All metrics cleared")
    }

    /** Clear metrics older than specified duration in milliseconds. */
    suspend fun clearOldMetrics(olderThanMs: Long) = mutex.withLock {
        val cutoff = Date(System.currentTimeMillis() - olderThanMs)

        val keysToRemove = latencyMetrics.entries
            .filter { (_, metric) -> metric.startTime == null || metric.startTime.before(cutoff) }
            .map { it.key }
        for (key in keysToRemove) {
            latencyMetrics.remove(key)
        }

        budgetViolations.removeAll { it.timestamp.before(cutoff) }
        resourceSnapshots.removeAll { it.timestamp.before(cutoff) }

        Log.i(TAG_INFERENCE, "Cleared metrics older than ${olderThanMs}ms")
    }

    // -------------------------------------------------------------------
    // Private - Trimming
    // -------------------------------------------------------------------

    /** Must be called under [mutex] lock. */
    private fun trimMetrics() {
        if (latencyMetrics.size > maxStoredMetrics) {
            val toRemove = latencyMetrics.size - maxStoredMetrics
            val oldestKeys = latencyMetrics.entries
                .sortedBy { it.value.startTime?.time ?: 0L }
                .take(toRemove)
                .map { it.key }
            for (key in oldestKeys) {
                latencyMetrics.remove(key)
            }
        }
    }

    /** Must be called under [mutex] lock. */
    private fun trimViolations() {
        if (budgetViolations.size > maxStoredViolations) {
            val toRemove = budgetViolations.size - maxStoredViolations
            repeat(toRemove) { budgetViolations.removeFirst() }
        }
    }

    /** Must be called under [mutex] lock. */
    private fun trimSnapshots() {
        if (resourceSnapshots.size > maxStoredSnapshots) {
            val toRemove = resourceSnapshots.size - maxStoredSnapshots
            repeat(toRemove) { resourceSnapshots.removeFirst() }
        }
    }
}

// ---------------------------------------------------------------------------
// Supporting Types
// ---------------------------------------------------------------------------

/** Latency metric for a single inference request. */
data class LatencyMetric(
    val requestId: String,
    val modelName: String?,
    val startTime: Date?,
    val endTime: Date?,
    val latencyMs: Double?
)

/** Budget violation record. */
data class BudgetViolationRecord(
    val timestamp: Date,
    val type: BudgetViolationType,
    val current: Double,
    val limit: Double,
    val severity: ViolationSeverity
)

/** Resource usage snapshot. */
data class ResourceSnapshot(
    val timestamp: Date,
    val memoryMb: Double,
    val batteryLevel: Double?,
    val thermalLevel: Int
)

/** Aggregated latency statistics. */
data class LatencyStats(
    val count: Int,
    val min: Double,
    val max: Double,
    val mean: Double,
    val p50: Double,
    val p95: Double,
    val p99: Double
) {
    override fun toString(): String =
        "LatencyStats(count=$count, min=${"%.2f".format(min)}ms, max=${"%.2f".format(max)}ms, " +
            "mean=${"%.2f".format(mean)}ms, p50=${"%.2f".format(p50)}ms, " +
            "p95=${"%.2f".format(p95)}ms, p99=${"%.2f".format(p99)}ms)"
}

/** Budget violation type. */
enum class BudgetViolationType(val value: String) {
    LATENCY("latency"),
    MEMORY("memory"),
    BATTERY("battery"),
    THERMAL("thermal")
}

/** Violation severity level. */
enum class ViolationSeverity(val value: String) {
    INFO("info"),
    WARNING("warning"),
    CRITICAL("critical")
}